en/ros2/px4_ros2_interface_lib.md

```diff
@@ -0,0 +1,478 @@
+  Here is where we can do our work and generate a new setpoint.
+
+After creating an instance of that mode, `mode->doRegister()` must be called which does the actual registration with the flight controller and returns `false` if it fails.
+In case a mode executor is used, `doRegister()` must be called on the mode executor, instead of for the mode.
+
+### Mode Executor Class Definition
+
+This section steps through an example of how to create a mode executor class.
+
+```cpp{1,4-5,9-16,20,33-57}
+class MyModeExecutor : public px4_ros2::ModeExecutorBase // [1]
+{
+public:
+  MyModeExecutor(rclcpp::Node & node, px4_ros2::ModeBase & owned_mode) // [2]
+  : ModeExecutorBase(node, px4_ros2::ModeExecutorBase::Settings{}, owned_mode),
+    _node(node)
+  { }
+
+  enum class State // [3]
+  {
+    Reset,
+    TakingOff,
+    MyMode,
+    RTL,
+    WaitUntilDisarmed,
+  };
+
+  void onActivate() override
+  {
+    runState(State::TakingOff, px4_ros2::Result::Success); // [4]
+  }
+
+  void onDeactivate(DeactivateReason reason) override { }
+
+  void runState(State state, px4_ros2::Result previous_result)
+  {
+    if (previous_result != px4_ros2::Result::Success) {
+      RCLCPP_ERROR(_node.get_logger(), "State %i: previous state failed: %s", (int)state,
+        resultToString(previous_result));
+      return;
+    }
+
+    switch (state) { // [5]
+      case State::Reset:
+        break;
+
+      case State::TakingOff:
+        takeoff([this](px4_ros2::Result result) {runState(State::MyMode, result);});
+        break;
+
+      case State::MyMode:
+        scheduleMode(
+          ownedMode().id(), [this](px4_ros2::Result result) {
+            runState(State::RTL, result);
+          });
+        break;
+
+      case State::RTL:
+        rtl([this](px4_ros2::Result result) {runState(State::WaitUntilDisarmed, result);});
+        break;
+
+      case State::WaitUntilDisarmed:
+        waitUntilDisarmed([this](px4_ros2::Result result) {
+            RCLCPP_INFO(_node.get_logger(), "All states complete (%s)", resultToString(result));
+          });
+        break;
+    }
+  }
+
+private:
+  rclcpp::Node & _node;
+};
+```
+
+- `[1]`: First we create a class that inherits from `px4_ros2::ModeExecutorBase`.
+- `[2]`: The constructor takes our custom mode that is associated with the executor and passes it to the constructor of `ModeExecutorBase`.
+- `[3]`: We define an enum for the states we want to run through.
+- `[4]`: `onActivate` gets called when the executor becomes active. At this point we can start to run through our states.
+  How you do this is up to you, in this example a method `runState` is used to execute the next state.
+- `[5]`: According to the state, we call a method from `ModeExecutorBase` that asynchronously runs the requested mode.
+  A callback is passed which is called upon completion.
+  This provides a `Result` argument which tells you whether the operation succeeded or not.
```

Comment:
My C++ is a bit rusty, but I think that  `run`, `takeoff`, `rtl`, `WaitUntilDisarmed` are functions that take a callback when they compete, and you are using these to drive the state machine - right?

Do we have a list of the available functions and explanation of their behaviour? Because I expect these have mappings to existing modes.

Further, we need to mention  `scheduleMode()` because I think that is how you run an arbitrary mode. Right?

Reply:
Yes and yes. We can mention it, but in general I think we should have generated API docs for these.

Reply:
@bkueng OK I spelled it out.

```suggestion
- `[1]`: First we create a class that inherits from `px4_ros2::ModeExecutorBase`.
- `[2]`: The constructor takes our custom mode that is associated with the executor and passes it to the constructor of `ModeExecutorBase`.
- `[3]`: We define an enum for the states we want to run through.
- `[4]`: `onActivate` gets called when the executor becomes active. At this point we can start to run through our states.
  How you do this is up to you, in this example a method `runState` is used to execute the next state.
- `[5]`: On switching to a state we call an asynchronous method from `ModeExecutorBase` to start the desired mode: `run`, `takeoff`, `rtl`, and so on.
  These methods are passed a function that is called on completion; the callback provides a `Result` argument that tells you whether the operation succeeded or not.
  The callback runs the next state on success.
- `[6]`: In the state handler for the mode executors "owned mode" we use the `scheduleMode()` method to start the mode.
  This follows the same callback pattern as the other methods/states.
```